### PR TITLE
Encapsulate state.OriginalTargets & add a lock

### PR DESCRIPTION
**Pull request by peterebden**
Saw a related panic that I've not seen before the other day. There is a data race around this guy; I had assumed it's benign but better to be safe.

### CODE REVIEW

File: src/core/state.go [hunk 1]
```diff
@@ -444,11 +447,9 @@ func (state *BuildState) AddOriginalTarget(label BuildLabel, addToList bool) {
...
 	}
 	if addToList {
-		// The sets of original targets are duplicated between states for all architectures,
-		// we must add it to all of them to ensure everything sees the same set.
-		for _, s := range state.progress.allStates {
-			s.OriginalTargets = append(s.OriginalTargets, label)
-		}
+		state.progress.originalTargetMutex.Lock()
+		state.progress.originalTargets = append(state.progress.originalTargets, label)
+		state.progress.originalTargetMutex.Unlock()
```
Comment (Tatskaari) [inline src/core/state.go:452]: Do we not care about other architectures anymore?

File: src/core/state.go [hunk 2]
```diff
@@ -578,7 +579,10 @@ func (state *BuildState) ExpandOriginalTargets() BuildLabels {
 // ExpandOriginalLabels expands any pseudo-labels (ie. :all, ... has already been resolved to a bunch :all targets)
 // from the set of original labels.
 func (state *BuildState) ExpandOriginalLabels() BuildLabels {
-	return state.ExpandLabels(state.OriginalTargets)
+	state.progress.originalTargetMutex.Lock()
+	targets := state.progress.originalTargets[:]
+	state.progress.originalTargetMutex.Unlock()
```
Comment (Tatskaari) [inline src/core/state.go:584]: nit: could use defer here

File: src/please.go [hunk 3]
```diff
@@ -468,8 +468,7 @@ var buildFunctions = map[string]func() int{
 	"gc": func() int {
 		success, state := runBuild(core.WholeGraph, false, false, true)
 		if success {
-			state.OriginalTargets = state.Config.Gc.Keep
-			gc.GarbageCollect(state, opts.Gc.Args.Targets, state.ExpandOriginalTargets(), state.Config.Gc.Keep, state.Config.Gc.KeepLabel,
+			gc.GarbageCollect(state, opts.Gc.Args.Targets, state.ExpandLabels(state.Config.Gc.Keep), state.Config.Gc.Keep, state.Config.Gc.KeepLabel,
```
Comment (Tatskaari) [inline src/please.go:471]: Just checking you meant to make this change. It's not obvious it's related.

File: src/core/state.go (hunk 1 above)
Comment (peterebden) [reply src/core/state.go:452]: the `progress` sub-structure is shared between all the states so it doesn't need the loop any more.

File: src/core/state.go (hunk 2 above)
Comment (peterebden) [reply src/core/state.go:584]: I don't want to hold the lock around ExpandLabels though

**Pull request merged**